ramips: fix ethernet bugs

mans0n opened the pull request:
ramips: 5.4: handle ERR_PTR properly
ramips: fix MAC address setup for Olimex RT5350F-OLinuXino devices
ramips: explicitly disable built-in switch for lan-only devices
ramips: remove default interface setup for some subtargets
ramips: rt305x: reset switch before probe
ramips: re-add set_preinit_iface script for rt305x and mt7620

mans0n committed:
ramips: 5.4: handle ERR_PTR properly

of_get_mac_address can return ERR_PTR since 5.2, so the return pointer should be
checked before used. Otherwise it might cause an oops during boot.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n committed:
ramips: fix MAC address setup for RT5350F-OLinuXino devices

Olimex RT5350F-OLinuXino devices do not have a default MAC address, and there is
nothing at the 0x4 offset in the factory partition. Using a local address, which
is randomly generated by the kernel, would be a better choice.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
The switch reset patch might be superseded by #2909.

mans0n commented:
Added switch disable patch.

mans0n commented:
Oops, I did not know NULL is also IS_ERR. Fixed now.

mans0n commented:
I found a case using eth0 mac address so I decided to move it to there.

hauke commented on target/linux/ramips/files/drivers/net/ethernet/ralink/mtk_eth_soc.c:
```diff
@@ -1401,7 +1401,11 @@ static int __init fe_init(struct net_device *dev)
 
 	mac_addr = of_get_mac_address(priv->dev->of_node);
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(5,2,0))
 	if (mac_addr)
```
You could use ´IS_ERR_OR_NULL()´ here, but I do not know if it is better.

mans0n replied:
That's great! I found some usages of `IS_ERR_OR_NULL` right after `of_get_mac_address` in the kernel, so I guess it's okay to use it here as well.

mans0n commented:
I'm trying to remove all of the "eth0.X" stuff in 02_network, but I'm not sure about how to deal with Dovado Tiny AC for example... I've never seen a device with such ethernet setup.
Until I find out a correct solution (or somebody else gives me a hint), I'd just add back the preinit script (in this PR).

mans0n commented:
- Replaced `IS_ERR` with `IS_ERR_OR_NULL`
- Added two patch
  - ramips: remove default interface setup for some subtargets
  - ramips: re-add set_preinit_iface script for rt305x and mt7620

If "re-add set_preinit_iface script" patch get accepted, then "reset switch before probe" patch can be dropped.

adrianschmutzler commented on target/linux/ramips/rt305x/base-files/etc/board.d/02_network:
```diff
@@ -230,6 +228,8 @@ ramips_setup_macs()
 	hilink,hlk-rm04|\
 	nexaira,bc2|\
+	olimex,rt5350f-olinuxino|\
+	olimex,rt5350f-olinuxino-evb|\
```
Hmm, don't really see why this particular "new" setup is chosen.

If factory 0x4 is empty, then the statement for ethernet in DTS should be removed in DTS, so a random address is actually generated. (For wmac, it will still use whatever is written in caldata?)

For 02_network, if there is no predefined MAC address, I do not see why we should change the lan_mac and assign a different wan_mac as well. I'd just have removed the devices here ...?

mans0n replied:
A random address (for ethernet) is actually generated, as the mac address at 0x4 is "ff:ff:ff:ff:ff:ff", and the kernel will not accept that. From what I've observed, the address generated by the kernel is always local, but we'd better play safe here by forcing local bit set.
By including the device in that case, the random address generated in the first boot will be saved in /etc/config/network and will be used in subsequent boots. For wmac, another random address is generated, but it is not saved in /etc/config/wireless so it will change on every reboot.
I thought the dts statement will have no harm, but I'll remove it if you like.

adrianschmutzler replied:
Hmm, I see why it works and what's the reason behind the 02_network entries now.

However, I don't really like the concept of relying on the calibration data to be all fs. An alternative option would be to set the local-mac-address to 00:00:00:00:00:00 or ff:ff:ff:ff:ff:ff manually in DTS, and thus trigger a random MAC address in any case.

And I'd be curious what the vendor did there and whether having a WAN port with separate address is actually justified.

Anyway, I admit this is all nitpicking, and your patch should address the problem properly as is....

mans0n replied:
> An alternative option would be to set the local-mac-address to 00:00:00:00:00:00 or ff:ff:ff:ff:ff:ff manually in DTS

That's a nice idea. One of the reason why I left the 0x4 statement in the dts was that users might write their mac address in factory partition for some reason. But probably we don't need to care that situation.

I don't see any particular reason to have a separate wan address here, but I prefer it to writing a new case only for this device. That's all.

mans0n replied:
It seems wmac address is not overridden by local-mac-address (nor mac-address). I'm not fond of using local-mac-address only for ethernet, so I'll just keep the current state.

adrianschmutzler commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -122,9 +122,6 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0:lan:2" "1:lan:1" "4:wan" "6@eth0"
 		;;
-	*)
```
This will result in the device becoming a one-port eventually (but a broken one), as there is a fallback to `ucidef_set_interface_lan eth0` hidden somewhere I don't remember (base-files package or something like that).

Rather than having this strange fallback nobody really knows about, I'd prefer to have the one-port setup as default case for the subtargets touched here.

mans0n replied:
For `ucidef_set_interfaces_lan_wan "eth0.1" "eth0.2"` case, the VLAN will be set according to the dts AFAIK, so the effect will be same as `ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" "wan"` in the default case of mt7621 02_network. (But it requires the preinit script, so it's broken now.)

> Rather than having this strange fallback nobody really knows about, I'd prefer to have the one-port setup as default case for the subtargets touched here.

Do you mean setting `ucidef_set_interface_lan eth0` explicitly as a default case?

adrianschmutzler replied:
Instead of just not having any default case here, I'd choose the
```
		ucidef_add_switch "switch0"
		ucidef_add_switch_attr "switch0" "enable" "false"
		ucidef_set_interface_lan "eth0"
		;;
```

mans0n replied:
Okay, understood.

adrianschmutzler commented:
I've picked patch 3 and will push it later this evening.

981213 commented on target/linux/ramips/rt3883/base-files/etc/board.d/02_network:
```diff
@@ -27,6 +27,8 @@ ramips_setup_interfaces()
 	loewe,wmdr-143n|\
 	omnima,hpm)
+		ucidef_add_switch "switch0"
+		ucidef_add_switch_attr "switch0" "enable" "false"
```
rt3883 doesn't have a built-in switch. This isn't needed.

mans0n replied:
Oh, sorry, I didn't know that. Then we should revert a1693bf626f8...

981213 commented on target/linux/ramips/rt288x/base-files/etc/board.d/02_network:
```diff
@@ -26,11 +26,10 @@ ramips_setup_interfaces()
 	buffalo,wli-tx4-ag300n|\
 	dlink,dap-1522-a1)
+		ucidef_add_switch "switch0"
+		ucidef_add_switch_attr "switch0" "enable" "false"
```
rt288x doesn't have a built-in switch. This isn't needed.

981213 commented on target/linux/ramips/mt7620/base-files/lib/preinit/07_set_preinit_iface_ramips:
```diff
@@ -0,0 +1,59 @@
+#
+# Copyright (C) 2013 OpenWrt.org
+#
+
```
NAK on these scripts. I'll make a commit to remove default case in 02_network.

mans0n replied:
Okay, that would be much better, thanks.

mans0n commented:
Removed all switch and default case patches. (Sorry for the wrong patch, I was not careful enough.)

mans0n commented:
As for the `local-mac-address`, I'll do more tests and push an amended patch.

adrianschmutzler commented:
I've reverted the patch.

981213 commented:
Merged into my staging tree at: https://git.openwrt.org/?p=openwrt/staging/981213.git
Thanks!
